pkg/flowaggregator/clickhouseclient/clickhouseclient.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clickhouseclient
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go"
+	"github.com/gammazero/deque"
+	ipfixentities "github.com/vmware/go-ipfix/pkg/entities"
+	"k8s.io/klog/v2"
+)
+
+const (
+	maxQueueSize   = 1 << 19 // 524288. ~500MB assuming 1KiB per record
+	commitInterval = 1 * time.Second
+	insertQuery    = "INSERT INTO antrea (sourceIP, destinationIP, sourcePort, destinationPort, sourcePodName, destinationPodName, bytesDelta, packetsDelta, flowEndSeconds) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)"
+)
+
+type ClickHouseExportProcess struct {
+	db    *sql.DB
+	dsn   string
+	deque *deque.Deque
+	mutex sync.RWMutex
+	// stopChan is the channel to receive stop message
+	stopChan chan bool
+}
+
+type ClickHouseInput struct {
+	Username string
+	Password string
+	Database string
+	DbURL    string
+	Debug    bool
+	Compress bool
+}
+
+func (ci *ClickHouseInput) getDataSourceName() (string, error) {
+	if len(ci.DbURL) == 0 || len(ci.Username) == 0 || len(ci.Password) == 0 {
+		return "", fmt.Errorf("URL, Username or Password missing for clickhouse DSN")
+	}
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("%s?username=%s&password=%s", ci.DbURL, ci.Username, ci.Password))
+
+	if len(ci.Database) > 0 {
+		sb.WriteString("&database=")
+		sb.WriteString(ci.Database)
+	}
+	if ci.Debug {
+		sb.WriteString("&debug=true")
+	}
+	if ci.Compress {
+		sb.WriteString("&compress=true")
+	}
+
+	return sb.String(), nil
+}
+
+type ClickHouseFlowRow struct {
```

Comment:
Thanks shawn working on this. Could you please update the code with all fields needed to be inserted into Clickhouse, according our current data schema: https://confluence.eng.vmware.com/display/~dyongming/Clickhouse+schema+and+Information+Elements+of+flow+record+in+long+term+arch#ClickhouseschemaandInformationElementsofflowrecordinlongtermarch-Basetable%22flows%22

Reply:
thanks. added. pls recheck.

Reply:
> Thanks shawn working on this. Could you please update the code with all fields needed to be inserted into Clickhouse, according our current data schema: https://confluence.eng.vmware.com/display/~dyongming/Clickhouse+schema+and+Information+Elements+of+flow+record+in+long+term+arch#ClickhouseschemaandInformationElementsofflowrecordinlongtermarch-Basetable%22flows%22

added. thanks!